redis/src/sentinel.rs

```diff
@@ -0,0 +1,795 @@
+    }
+
+    /// Get a list of all masters (using the command SENTINEL MASTERS) from the
+    /// sentinels (we'll return the result of the first sentinel who responds to that
+    /// command without errors).
+    fn get_sentinel_masters(&mut self) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(sentinel_masters_cmd())
+    }
+
+    fn get_sentinel_replicas(
+        &mut self,
+        service_name: &str,
+    ) -> RedisResult<Vec<HashMap<String, String>>> {
+        self.try_all_sentinels(sentinel_replicas_cmd(service_name))
+    }
+
+    fn find_master_address(
+        &mut self,
+        service_name: &str,
+        node_connection_info: &SentinelNodeConnectionInfo,
+    ) -> RedisResult<ConnectionInfo> {
+        let masters = self.get_sentinel_masters()?;
+        find_valid_master(masters, service_name, node_connection_info)
+    }
+
+    fn find_valid_replica_addresses(
+        &mut self,
+        service_name: &str,
+        node_connection_info: &SentinelNodeConnectionInfo,
+    ) -> RedisResult<Vec<ConnectionInfo>> {
+        let replicas = self.get_sentinel_replicas(service_name)?;
+        Ok(get_valid_replicas_addresses(replicas, node_connection_info))
+    }
+
+    /// Determines the masters address for the given name, and returns a client for that
+    /// master.
+    ///
+    /// To determine the master we ask the sentinels one by one until one of
+    /// then correctly responds to the SENTINEL MASTERS command, and then we check to
+    /// see if there is a master with the given name; if there is, we create a
+    /// client with that address (ip and port) and the extra information from the
+    /// `node_connection_info` parameter.
+    pub fn master_for(
+        &mut self,
+        service_name: &str,
+        node_connection_info: Option<&SentinelNodeConnectionInfo>,
+    ) -> RedisResult<Client> {
+        let connection_info =
+            self.find_master_address(service_name, node_connection_info.unwrap_or_default())?;
+        Client::open(connection_info)
+    }
+
+    /// Connects to the first available replica of the given master name.
+    ///
+    /// After listing the valid replicas (those without a sdown or odown flag) for the
+    /// given master name, we start trying to connect to then one by one, starting at a
+    /// random index (simple random derived from the nanoseconds part of the current
+    /// unix timestamp).
+    pub fn replica_for(
+        &mut self,
+        service_name: &str,
+        node_connection_info: Option<&SentinelNodeConnectionInfo>,
+    ) -> RedisResult<Client> {
+        let addresses = self
+            .find_valid_replica_addresses(service_name, node_connection_info.unwrap_or_default())?;
+        let start_index = random_replica_index(addresses.len());
+        try_connect_to_first_replica(&addresses, start_index)
+    }
+
+    /// Connects to the first available replica of the given master name, starting in a
```

Comment:
This doc is a bit vague - what is the "first available replica", and what does the index index?

Specifically, there seems to be a hidden assumption that the list of replicas is the same between calls to the function, which is wrong if any replica was replaced.
I don't have a clearer phrasing off the top of my head, and I'm also not 100% sure that this is a meaningful problem, so my request here is just to see if you have better phrasing. If not, I believe this is fine.

Reply:
I think this "first available replica" comes from a previous state of the code, and is related to using the function `try_connect_to_first_replica`. I tried to rewrite this whole description to try and make it clearer!